File: go/TwoPointers/11_container_with_most_water/11_container_with_most_water.go
package containerwithmostwater

func maxArea(height []int) int {
	var maxArea int
	leftPointer := 0
	rightPointer := len(height) - 1
	for leftPointer < rightPointer {
		area := (rightPointer - leftPointer) * min(height[leftPointer], height[rightPointer])
		maxArea = max(area, maxArea)
		if height[leftPointer] < height[rightPointer] {
			leftPointer++
		} else {
			rightPointer--
		}
	}
	return maxArea
}

func min(i, j int) int {
	if i < j {
		return i
	}
	return j
}

func max(i, j int) int {
	if i > j {
		return i
	}
	return j
}

File: rust/02_two_pointers/11_container_with_most_water.rs
impl Solution {
    pub fn max_area(height: Vec<i32>) -> i32 {
        let (mut l, mut r) = (0, height.len() - 1);
        
        let mut max = 0;
        
        while (l < r){
            let (lh, rh) = (height[l], height[r]);
            let h = lh.min(rh);
            
            let d = (r - l) as i32;
            let area = d * h;
            
            if area > max{
                max = area;
            }
            
            if rh< lh{
                while r > 0 && height[r] <= rh{
                    r-=1;
                }
            }else{
                while l < height.len() && height[l] <= lh{
                    l+=1;
                }
            }
        }
        
        max
    }
}